cleverhans/experimental/certification/optimization.py

```diff
@@ -216,7 +313,7 @@ def run_optimization(self):
       True if certificate is found
       False otherwise
     """
-    self.prepare_one_step()
+    self.prepare_for_optimization()
```

Comment:
What I meant is to move call to `prepare_for_optimization` into init function of `Optimization` class. This won't change how user interact with this class (`prepare_for_optimization` will be called either when class is created or when `run_optimization` is called), in both cases user will do following:
```python
opt = Optimization(...)
result =  opt.run_optimization()
```

The main reason why I was thinking to put call to self.prepare_for_optimization into `__init__` function because you already creating some ops in the `__init__` function (creating bunch of placeholders and call to `project_dual`).
It's not super important, but generally it's nice for readability to keep all the code which construct ops in one place.
Otherwise reader have to look at multiple methods to figure our what ops are being constructed.

Back to your concern, I would suggest you to think how users are supposed to use Optimization class:

* is it a one time object which is created every time you run optimization? If answer is yes, then it does not matter too much where exactly you put call to `self.prepare_for_optimization`
* Are you going to reuse `Optimization` class and call `run_optimization` several times? if yes, then probably it's a good idea to avoid constructing ops in the `run_optimization` method. Also in this case it worth thinking about how subsequent calls to `run_optimization` would differ from each other (i.e. right now no arguments are passed to `run_optimization`, so even if you call it several times, each subsequent call will do the same thing)

I'll let you think and make a decision regarding what to do